packages/firestore/src/util/assert.ts

```diff
@@ -18,32 +18,54 @@
 import { SDK_VERSION } from '../core/version';
 import { logError } from './log';
 
+/**
+ * Fails if the given assertion condition is false, throwing an Error with the
+ * given message if it did.
+ */
+export function hardAssert(
+  assertion: boolean,
+  message: string
+): asserts assertion {
+  if (!assertion) {
+    // Log the failure in addition to throw an exception, just in case the
+    // exception is swallowed.
+    const errorMessage =
+      `FIRESTORE (${SDK_VERSION}) INTERNAL ASSERTION FAILED: ` + message;
+    logError(errorMessage);
+
+    // NOTE: We don't use FirestoreError here because these are internal failures
+    // that cannot be handled by the user. (Also it would create a circular
+    // dependency between the error and assert modules which doesn't work.)
+    throw new Error(errorMessage);
+  }
+}
+
 /**
  * Unconditionally fails, throwing an Error with the given message.
  *
+ * The code of callsites invoking this function are stripped out in production
```

Comment:
Apologies--I totally missed that you're stripping out calls to `fail()` as well. There are a few cases where data outside our control results in a failure that would be unhandled. File and line numbers taken from 761d0d3a48175fdcbee5b43feb66d89c4b6567a7.

  * in database.ts:676 (`Transattion.get`): `if (!docs || docs.length !== 1)` -- this seems like it could straightforwardly be rewritten as a `hardAssert`
  * in user_data_reader:356 (`parseMergeData`): it seems like this may be validating user input? Possibly this should be converted to a validation error (or possibly this was validated elsewhere)
  * in query.ts:494 (`Operator.fromString`): the default case seems like it might be validating user input. Oh wait, this is validated at database.ts:1459. This one's OK, I think.
  * in encoded_resource_path.ts:142, 170 (`decodeResourcePath`): this is decoding a resource path from IndexedDB
  * in indexeddb_mutation_queue.ts:348 (`getAllMutationBatchesAffectingDocumentKey`): dangling document-references could reflect corruption in IndexedDB
  * Same on line 480 (`lookupMutationBatches`)
  * local_serializer.ts:67: (`fromDbRemoteDocument`): borderline case, but your `fromFoo` in `serializer.ts` seemed mostly to go toward `hardAssert` so I thought I'd flag
  * values.ts:380: (`canonifyValue`): This is how we process input from the network, so it's possible to get unexpected values here. The other `fail` calls in this file are all default cases for `TypeOrder` so I think they're fine.
  * webchannel_connection.ts:170: RPC error code has a remote source

Actually I'm going to stop--there seem like more and we should at least discuss this. A `hardFail`/`debugFail` split would definitely address this (or just leaving all `fail` calls in too, though that cuts into size savings).

Many of these seem like they could be reformulated in terms of the error handling that's in place, but that seems out of scope for this change.

Reply:
I reverted back the documentation change for `fail`. Let's tackle this one API at a time. If we want to keep fail() but reduce the size impact we can remove the error message, but first I would like to look at the actual size impact of the messages.

So, for now, this PR differentiates between the different assertions, and then I will update the Transformer PR to not stip `fail()`.